### PR TITLE
#0: TG-Llama3-70b - Add compilation step to demo

### PR DESCRIPTION
**Pull request by djordje-tt**
### Problem description
- Demo does not compile nor prefill nor decode so 1st prefill run is very slow and after prefill is done 1st decode iteration is super slow and there is 1min delay between getting 1st token from prefill and next token from decode.
- Time to first token also added to output statistics.

### What's changed
There is 1-layer compilation added at the beginning of demo as warmup iteration for both prefill and decode. Now there is no delay between prefill and decode. Also, there is no need to run attention for every multiple_of_32 since k_chunk_size is 512 and kernel is compiled in compilation and is reused later no matter which cur_pos is for decode.

### CODE REVIEW

File: models/demos/tg/llama3_70b/demo/demo.py [hunk 1]
```diff
@@ -76,6 +76,56 @@ def construct_arg(**kwargs):
...
+
+
+def demo_warmup(args):
+    # Copy and modify arguments
+    model_args = replace(args.model, num_layers=1)
+    tt_args = replace(args.tt)
+    data_args = replace(args.data, max_output_tokens=2)
+
+    # Skip if model_implementation is not tt
+    if model_args.implementation != "tt":
```
Comment (johanna-rock-tt) [inline models/demos/tg/llama3_70b/demo/demo.py:103]: Why not do this at the very start of this method? Do we want to update the model args (e.g. num_layers=1) if we don't run tt version?

File: models/demos/tg/llama3_70b/demo/demo.py [hunk 2]
```diff
@@ -76,6 +76,56 @@ def construct_arg(**kwargs):
...
+        return
+
+    generator = build_generator(model_args, tt_args)
+
+    # Load the model and tokenizer
+    model, tokenizer = generator.model, generator.tokenizer
+
+    tokenized, prompts = load_prompts_file(model_args, data_args, tokenizer)
```
Comment (johanna-rock-tt) [inline models/demos/tg/llama3_70b/demo/demo.py:111]: Why don't we just use random inputs with the right shapes here? I don't think we need to load the prompts file and get the correctly sizes promps out of there.

File: models/demos/tg/llama3_70b/demo/demo.py [hunk 3]
```diff
@@ -255,14 +307,21 @@ def run_decode(
...
         next_logits = logits[:, -1, :]  # batch, vocab of last token
         next_token = sampling_func(next_logits)
 
         tokens, eos_reached, prev_pos = prepare_next_input(
             tokenizer, tokens, input_text_mask, finished_mask, prompt_lens, cur_pos, next_token
         )
         latencies.append(time() - start)
-
+        # If compilation run for 2 iterations (prefill+decode)
+        if compilation and min_prompt_len + 2 == cur_pos:
```
Comment (johanna-rock-tt) [inline models/demos/tg/llama3_70b/demo/demo.py:321]: Wouldn't it be enough to just call model.forward and prepare_next_input in a loop for warmup? Or even push random inputs? I think it would be cleaner if we wouldn't re-use run_decode and then continue/break dependent on compilation and what position we are in.

File: models/demos/tg/llama3_70b/demo/demo.py [hunk 4]
```diff
@@ -255,14 +307,21 @@ def run_decode(
...
+        if cur_pos == min_prompt_len:
+            total_time_to_first_token_end = time()
+
         next_logits = logits[:, -1, :]  # batch, vocab of last token
         next_token = sampling_func(next_logits)
 
         tokens, eos_reached, prev_pos = prepare_next_input(
             tokenizer, tokens, input_text_mask, finished_mask, prompt_lens, cur_pos, next_token
         )
         latencies.append(time() - start)
```
Comment (johanna-rock-tt) [inline models/demos/tg/llama3_70b/demo/demo.py:319]:
```
if cur_pos == min_prompt_len:
    total_time_to_first_token = time() - start
```
?

File: models/demos/tg/llama3_70b/demo/demo.py (hunk 2 above)
Comment (djordje-tt) [reply models/demos/tg/llama3_70b/demo/demo.py:111]: Since Jack added option for k_chunk_size then if we use it we might not hit precompiled kernel every time. With real shapes we are using same examples. Loading file is super fast and I don't see much of a benefit with using random inputs over data from files.

File: models/demos/tg/llama3_70b/demo/demo.py (hunk 3 above)
Comment (djordje-tt) [reply models/demos/tg/llama3_70b/demo/demo.py:321]: I believe doing everything the same what run_decode already does in warmup step would make code more complex and when we can reuse run_decode I believe we shouldn't add more code to there. I agree that if statements for compilation make run_decode less clean but I wouldn't say it's less understandable or more complex. Reusing it for compilation step clearly shows that we are doing exactly the same thing as in forward pass later.

File: models/demos/tg/llama3_70b/demo/demo.py (hunk 4 above)
Comment (djordje-tt) [reply models/demos/tg/llama3_70b/demo/demo.py:319]: Prefill stage is done in one iteration of surrounding for loop. model.forward calls prefill_forward which calls prefill for 32 users and total_time_to_first_token is then averaged with num_users later in the code.

File: models/demos/tg/llama3_70b/demo/demo.py (hunk 1 above)
Comment (djordje-tt) [reply models/demos/tg/llama3_70b/demo/demo.py]: Agreed. Modified.

**Review by johanna-rock-tt — APPROVED** [review]
